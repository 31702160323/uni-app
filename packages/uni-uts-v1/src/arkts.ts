import path from 'path'
import fs from 'fs-extra'
import type { UTSBundleOptions } from '@dcloudio/uts'
import { formatUniProviderName, getUTSCompiler } from './utils'
import type { CompileResult } from '.'

interface ArkTSCompilerOptions {
  isX?: boolean
  isExtApi?: boolean
  transform?: {
    uniExtApiProviderName?: string
    uniExtApiProviderService?: string
    uniExtApiProviderServicePlugin?: string
  }
}

type AutoImportOptions = Record<string, [string, (string | undefined)?][]>

export function mergeArkTSAutoImports(
  base: AutoImportOptions,
  ext: AutoImportOptions
): AutoImportOptions {
  const keys = new Set([...Object.keys(base), ...Object.keys(ext)])
  const result: AutoImportOptions = {}
  for (const key of keys) {
    const baseImports = base[key] || []
    const extImports = ext[key] || []
    result[key] = [...baseImports, ...extImports]
  }
  return result
}

export function getArkTSAutoImports(): AutoImportOptions {
  return mergeArkTSAutoImports(
    {
      '@dcloudio/uni-app-runtime': [
        ['defineAsyncApi'],
        ['defineSyncApi'],
        ['defineTaskApi'],
        ['defineOnApi'],
        ['defineOffApi'],
        ['getUniProvider'],
        ['getUniProviders'],
        ['string'],
        ['AsyncApiSuccessResult'],
        ['AsyncApiResult'],
        ['ApiExecutor'],
        ['ComponentInternalInstance'],
        ['ComponentPublicInstance'],
        ['IUniError'],
        ['ProtocolOptions'],
        ['ApiOptions'],
        ['ApiError'],
        ['UniError'],
        ['UniProvider'],
        ['uni'],
        ['IUTSObject'],
        ['UTSObject'],
        ['UTSJSONObject'],
      ],
    },
    require('../lib/arkts/ext-api-export.json')
  )
}
export async function compileArkTS(
  pluginDir: string,
  { isExtApi, transform }: ArkTSCompilerOptions
): Promise<CompileResult | void> {
  const filename = resolveAppHarmonyIndexFile(pluginDir)
  if (!filename) {
    return
  }

  const { bundle, UTSTarget } = getUTSCompiler()
  const pluginId = path.basename(pluginDir)
  const inputDir = process.env.UNI_INPUT_DIR
  const outputUniModuleDir = resolveAppHarmonyUniModuleDir(pluginId)

  const autoImportExternals = getArkTSAutoImports()
  if (transform && transform.uniExtApiProviderService) {
    autoImportExternals['@dcloudio/uni-app-runtime'].push([
      formatUniProviderName(transform.uniExtApiProviderService),
    ])
  }

  const buildOptions: UTSBundleOptions = {
    hbxVersion: process.env.HX_Version || process.env.UNI_COMPILER_VERSION,
    input: {
      root: inputDir,
      filename,
      paths: {
        '@dcloudio/uni-runtime': '@dcloudio/uni-app-runtime',
      },
      parseOptions: {
        tsx: true,
        noEarlyErrors: true,
        allowComplexUnionType: true,
        allowTsLitType: true,
      },
    },
    output: {
      outDir: outputUniModuleDir,
      outFilename: 'utssdk/app-harmony/index.ets',
      package: '',
      imports: [],
      sourceMap: false,
      extname: '.ets',
      logFilename: false,
      isPlugin: true,
      transform: {
        autoImportExternals,
      },
      treeshake: {
        noSideEffects: true,
      },
    },
  }
  const configFilePath = path.resolve(
    pluginDir,
    'utssdk/app-harmony/config.json'
  )

  const harmonyPackageName = '@uni_modules/' + pluginId
  const harmonyModuleName = harmonyPackageName
    .replace(/@/g, '')
    .replace(/\//g, '__')
    .replace(/-/g, '_')

  // generate oh-package.json5
  const ohPackageJson: Record<string, any> = {
    name: harmonyPackageName,
    version: '1.0.0',
    description: '',
    main: 'utssdk/app-harmony/index.ets',
    author: '',
    license: '',
    dependencies: {},
  }
  if (fs.existsSync(configFilePath)) {
    const config = fs.readJSONSync(configFilePath)
    ohPackageJson.dependencies = config.dependencies
    ohPackageJson.dynamicDependencies = config.dynamicDependencies
    ohPackageJson.devDependencies = config.devDependencies
  }
  fs.outputJSONSync(
    path.resolve(outputUniModuleDir, 'oh-package.json5'),
    ohPackageJson,
    {
      spaces: 2,
    }
  )

  // copy resources
  const resourcesDir = path.resolve(pluginDir, 'utssdk/app-harmony/resources')
  if (fs.existsSync(resourcesDir)) {
    fs.copySync(
      resourcesDir,
      path.resolve(outputUniModuleDir, 'src/main/resources')
    )
  }

  // TODO 以下文件用户可定制

  // src/main/module.json5
  fs.outputJSONSync(
    path.resolve(outputUniModuleDir, 'src/main/module.json5'),
    {
      module: {
        name: harmonyModuleName,
        type: 'har',
        deviceTypes: ['default', 'tablet', '2in1'],
      },
    },
    {
      spaces: 2,
    }
  )

  // build-profile.json5
  fs.outputJSONSync(
    path.resolve(outputUniModuleDir, 'build-profile.json5'),
    {
      apiType: 'stageMode',
      buildOption: {},
      buildOptionSet: [],
      targets: [
        {
          name: 'default',
        },
      ],
    },
    {
      spaces: 2,
    }
  )

  // hvigorfile.ts
  fs.outputFileSync(
    path.resolve(outputUniModuleDir, 'hvigorfile.ts'),
    `import { harTasks } from '@ohos/hvigor-ohos-plugin';

export default {
    system: harTasks,  /* Built-in plugin of Hvigor. It cannot be modified. */
    plugins:[]         /* Custom plugin to extend the functionality of Hvigor. */
}`
  )

  const result = await bundle(UTSTarget.ARKTS, buildOptions)
  const deps: string[] = [filename]
  if (process.env.NODE_ENV === 'development') {
    if (result.deps) {
      deps.push(...result.deps)
    }
  }
  return {
    code: requireUTSPluginCode(pluginId, !!isExtApi),
    deps,
    encrypt: true,
    dir: outputUniModuleDir,
    inject_apis: [],
    scoped_slots: [],
  }
}

function requireUTSPluginCode(pluginId: string, isExtApi: boolean) {
  if (isExtApi) {
    return `export default uni`
  }
  return `export default uni.requireUTSPlugin('uni_modules/${pluginId}')`
}

function resolveAppHarmonyIndexFile(pluginDir: string) {
  let indexFile = path.resolve(pluginDir, 'utssdk/app-harmony/index.uts')
  if (fs.existsSync(indexFile)) {
    return indexFile
  }
  indexFile = path.resolve(pluginDir, 'utssdk/index.uts')
  if (fs.existsSync(indexFile)) {
    return indexFile
  }
}

export function resolveAppHarmonyUniModulesRootDir() {
  if (process.env.UNI_APP_HARMONY_PROJECT_PATH) {
    return path.resolve(process.env.UNI_APP_HARMONY_PROJECT_PATH, 'uni_modules')
  }
  return path.resolve(process.env.UNI_OUTPUT_DIR, 'uni_modules')
}

export function resolveAppHarmonyUniModuleDir(pluginId: string) {
  return path.resolve(resolveAppHarmonyUniModulesRootDir(), pluginId)
}

export function resolveAppHarmonyUniModulesEntryDir() {
  if (process.env.UNI_APP_HARMONY_PROJECT_PATH) {
    return path.resolve(
      process.env.UNI_APP_HARMONY_PROJECT_PATH,
      'entry/src/main/ets/uni_modules'
    )
  }
  return path.resolve(process.env.UNI_OUTPUT_DIR, 'uni_modules')
}
