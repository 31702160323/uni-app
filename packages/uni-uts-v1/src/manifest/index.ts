import { existsSync } from 'fs'
import { join } from 'path'
import {
  checkManifest,
  hasCustomResources,
  isCustomResources,
  resolveManifestJson,
  resolvePluginAndroidFiles,
} from './manifest'
import {
  APP_PLATFORM,
  CheckOptions,
  CheckResult,
  customResourceChangedTips,
  customResourceTips,
} from './utils'

interface PlatformOptions {
  customRes: string[]
  cacheFile: false | string
}

const ANDROID_CUSTOM_RES = [
  'app-android/assets/',
  'app-android/libs/',
  'app-android/res/',
  'app-android/AndroidManifest.xml',
  'app-android/config.json',
]

const IOS_CUSTOM_RES = [
  'app-ios/Frameworks/',
  'app-ios/Resources/',
  'app-ios/Info.plist',
  'app-ios/config.json',
]

export function checkKotlinCompile(
  playground: typeof process.env.HX_USE_BASE_TYPE,
  options: CheckOptions
) {
  return checkCompile('app-android', playground, options)
}

export function checkSwiftCompile(
  playground: typeof process.env.HX_USE_BASE_TYPE,
  options: CheckOptions
) {
  return checkCompile('app-ios', playground, options)
}

function checkCompile(
  platform: APP_PLATFORM,
  playground: typeof process.env.HX_USE_BASE_TYPE,
  options: CheckOptions
) {
  const platformOptions: PlatformOptions = {
    customRes: platform === 'app-android' ? ANDROID_CUSTOM_RES : IOS_CUSTOM_RES,
    cacheFile:
      platform === 'app-android'
        ? resolveDexCacheFile(options.pluginRelativeDir, options.outputDir)
        : false,
  }
  if (playground === 'standard') {
    return checkWithPlayground('standard', options, platformOptions)
  }
  return checkWithPlayground('custom', options, platformOptions)
}

async function checkWithPlayground(
  type: typeof process.env.HX_USE_BASE_TYPE,
  {
    id,
    env,
    cacheDir,
    pluginDir,
    pluginRelativeDir,
    is_uni_modules,
  }: CheckOptions,
  { customRes, cacheFile }: PlatformOptions
): Promise<CheckResult> {
  // 第一步：获取所有文件列表
  const files = await resolvePluginAndroidFiles(pluginDir, is_uni_modules)
  let tips = ''
  // 标准基座检查是否包含原生资源/配置
  if (type === 'standard' && hasCustomResources(files, customRes)) {
    tips = customResourceTips(id)
  }
  // 第二步：检查 dex 文件是否存在
  if (cacheFile !== false && !cacheFile) {
    return { expired: true, tips, cacheFile: '' }
  }
  // 第三步：获取当前插件缓存文件信息
  const manifest = resolveManifestJson(
    'app-android',
    pluginRelativeDir,
    cacheDir
  )
  if (!manifest) {
    return { expired: true, tips, cacheFile: '' }
  }
  // 第四步：检查文件变更
  const res = await checkManifest(manifest, { env, files, pluginDir })
  // 自定义基座检查原生资源/配置是否发生变化
  if (type === 'custom' && typeof res === 'string') {
    if (isCustomResources(res, customRes)) {
      tips = customResourceChangedTips(id)
    }
  }
  return {
    expired: res !== true,
    tips,
    cacheFile,
  }
}

function resolveDexCacheFile(pluginRelativeDir: string, outputDir: string) {
  const file = join(outputDir, '../.uts/dex', pluginRelativeDir, 'classes.dex')
  return (existsSync(file) && file) || ''
}
