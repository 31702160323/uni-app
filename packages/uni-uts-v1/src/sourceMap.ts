import { existsSync, readFileSync, statSync } from 'fs'
import { basename, dirname, join, relative } from 'path'
import {
  BasicSourceMapConsumer,
  IndexedSourceMapConsumer,
  NullableMappedPosition,
  NullablePosition,
  SourceMapConsumer,
} from 'source-map'
import { isWindows } from './shared'

const EXTNAME = {
  kotlin: '.kt',
  swift: '.swift',
}

const PLATFORM_DIR = {
  kotlin: 'app-android',
  swift: 'app-ios',
}

export function resolveUtsPluginSourceMapFile(
  target: 'kotlin' | 'swift',
  filename: string,
  inputDir: string,
  outputDir: string
) {
  const pluginDir = resolvePluginDir(inputDir, outputDir, filename)
  const is_uni_modules = basename(dirname(pluginDir)) === 'uni_modules'
  const sourceMapFile = join(
    join(outputDir, '../.sourcemap/app'),
    relative(inputDir, pluginDir),
    is_uni_modules ? 'utssdk' : '',
    PLATFORM_DIR[target],
    `index${EXTNAME[target]}.map`
  )
  if (!existsSync(sourceMapFile)) {
    throw `${sourceMapFile} not found`
  }
  return sourceMapFile
}

function resolvePluginDir(
  inputDir: string,
  outputDir: string,
  filename: string
) {
  // 目标文件是编译后 kt 或 swift
  if (filename.startsWith(outputDir)) {
    const relativePath = relative(outputDir, filename)
    // uni_modules/test-uts
    if (relativePath.startsWith('uni_modules')) {
      return join(inputDir, join(relativePath, '../../..'))
    }
    // utssdk/test-uts
    return join(inputDir, join(relativePath, '../..'))
  } else if (filename.startsWith(inputDir)) {
    let parent = dirname(filename)
    const utssdkDir = join(inputDir, 'utssdk')
    const uniModulesDir = join(inputDir, 'uni_modules')
    while (parent) {
      const dir = dirname(parent)
      if (dir === utssdkDir || dir === uniModulesDir) {
        return parent
      }
      parent = dir
    }
    throw `${filename} is not a uts plugin file`
  } else {
    throw `${filename} is not in ${inputDir} or ${outputDir}`
  }
}

interface PositionFor {
  sourceMapFile: string
  filename: string
  line: number
  column: number
}

const consumers: Record<
  string,
  { time: number; consumer: BasicSourceMapConsumer | IndexedSourceMapConsumer }
> = {}

/**
 * 根据源码文件名、行号、列号，返回生成后文件、行号、列号（根据 uts 文件返回 kt|swift 文件）
 * @param originalPosition
 * @returns
 */
export function generatedPositionFor({
  sourceMapFile,
  filename,
  line,
  column,
}: PositionFor): Promise<NullablePosition> {
  return resolveSourceMapConsumer(sourceMapFile).then((consumer) => {
    return consumer.generatedPositionFor({
      source: isWindows ? `\\\\?\\` : '' + filename,
      line,
      column,
    })
  })
}

/**
 * 根据生成后的文件名、行号、列号，返回源码文件、行号、列号（根据 kt|swift 文件返回 uts 文件）
 * @param generatedPosition
 * @returns
 */
export function originalPositionFor(
  generatedPosition: PositionFor
): Promise<NullableMappedPosition> {
  return resolveSourceMapConsumer(generatedPosition.sourceMapFile).then(
    (consumer) => {
      return consumer.originalPositionFor({
        line: generatedPosition.line,
        column: generatedPosition.column,
      })
    }
  )
}

async function resolveSourceMapConsumer(sourceMapFile: string) {
  const stats = statSync(sourceMapFile)
  if (!stats.isFile()) {
    throw `${sourceMapFile} is not a file`
  }
  const cache = consumers[sourceMapFile]
  if (!cache || cache.time !== stats.mtimeMs) {
    consumers[sourceMapFile] = {
      time: stats.mtimeMs,
      consumer: await new SourceMapConsumer(
        readFileSync(sourceMapFile, 'utf8')
      ),
    }
  }
  return consumers[sourceMapFile].consumer
}
