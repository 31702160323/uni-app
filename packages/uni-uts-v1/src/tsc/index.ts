export { uts2js } from './javascript'
export { runUTS2KotlinDev } from './kotlin'

import type { CompilerOptions, EmitResult } from 'typescript'

export interface UTSEmitDeclarationOptions {
  typescript: typeof import('typescript')
  inputDir: string
  rootFiles: string[]
  compilerOptions: CompilerOptions
}

export function emitDeclaration(
  options: UTSEmitDeclarationOptions
): EmitResult {
  return require('../../../lib/uts').emitDeclaration(options)
}
