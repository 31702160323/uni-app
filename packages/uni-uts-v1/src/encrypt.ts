import path, { join, relative } from 'path'
import fs from 'fs-extra'
import { APP_PLATFORM } from './manifest/utils'

export function isEncrypt(pluginDir: string) {
  return fs.existsSync(path.resolve(pluginDir, 'encrypt'))
}

export async function compileEncrypt(pluginDir: string) {
  const inputDir = process.env.UNI_INPUT_DIR
  const outputDir = process.env.UNI_OUTPUT_DIR
  const utsPlatform = process.env.UNI_UTS_PLATFORM as APP_PLATFORM
  const pluginRelativeDir = relative(inputDir, pluginDir)
  if (process.env.NODE_ENV !== 'development') {
    // 复制插件目录
    fs.copySync(pluginDir, join(outputDir, pluginRelativeDir))
  } else {
    // 读取缓存目录的 js code
    const cacheDir = process.env.HX_DEPENDENCIES_DIR!
    const indexJsPath = resolveJsCodeCacheFilename(
      utsPlatform,
      cacheDir,
      pluginRelativeDir
    )
    if (!fs.existsSync(indexJsPath)) {
      return console.error(
        `uts插件[${path.dirname(pluginDir)}]不存在，请重新云打包`
      )
    }
    return {
      code: fs.readFileSync(indexJsPath, 'utf-8'),
      deps: [] as string[],
    }
  }
  return
}

export function resolveJsCodeCacheFilename(
  platform: APP_PLATFORM,
  cacheDir: string,
  pluginRelativeDir: string
) {
  return join(cacheDir, platform, 'uts', pluginRelativeDir, 'index.js')
}
