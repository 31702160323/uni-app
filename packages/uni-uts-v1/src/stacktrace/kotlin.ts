import path from 'path'
import fs from 'fs-extra'
import { relative } from '../utils'
import { originalPositionFor } from '../sourceMap'
import { generateCodeFrame } from './utils'

export interface MessageSourceLocation {
  type: 'exception' | 'error' | 'warning' | 'info' | 'logging' | 'output'
  message: string
  file?: string
  line?: number
  column?: number
  code?: string
}

interface GenerateCodeFrameOptions {
  inputDir: string
  sourceMapDir: string
  replaceTabsWithSpace?: boolean
  format: (msg: MessageSourceLocation) => string
}

export function hbuilderFormatter(m: MessageSourceLocation) {
  const msgs: string[] = []
  let msg = m.type + ': ' + m.message
  if (m.type === 'warning') {
    // 忽略部分警告
    if (msg.includes(`Classpath entry points to a non-existent location:`)) {
      return ''
    }
    msg
      .replace(/\r\n/g, '\n')
      .split('\n')
      .forEach((m) => {
        msgs.push('\u200B' + m + '\u200B')
      })
  } else if (m.type === 'error' || m.type === 'exception') {
    msg
      .replace(/\r\n/g, '\n')
      .split('\n')
      .forEach((m) => {
        msgs.push('\u200C' + m + '\u200C')
      })
  } else {
    msgs.push(msg)
  }
  if (m.file) {
    if (m.file.includes('?')) {
      ;[m.file] = m.file.split('?')
    }
    msgs.push(`at ${m.file}:${m.line}:${m.column}`)
  }
  if (m.code) {
    msgs.push(m.code)
  }
  return msgs.join('\n')
}

export async function parseUTSKotlinStacktrace(
  messages: MessageSourceLocation[],
  options: GenerateCodeFrameOptions
) {
  if (typeof messages === 'string') {
    try {
      messages = JSON.parse(messages)
    } catch (e) {}
  }
  const msgs: string[] = []
  if (Array.isArray(messages) && messages.length) {
    function resolveSourceMapFile(file: string) {
      const sourceMapFile = path.resolve(
        options.sourceMapDir,
        relative(file, options.inputDir) + '.map'
      )
      if (fs.existsSync(sourceMapFile)) {
        return sourceMapFile
      }
    }
    for (const m of messages) {
      if (m.file) {
        const sourceMapFile = resolveSourceMapFile(m.file)
        if (sourceMapFile) {
          const originalPosition = await originalPositionFor({
            sourceMapFile,
            line: m.line!,
            column: m.column!,
            withSourceContent: true,
          })

          if (originalPosition.source && originalPosition.sourceContent) {
            m.file = originalPosition.source.split('?')[0]
            if (originalPosition.line !== null) {
              m.line = originalPosition.line
            }
            if (originalPosition.column !== null) {
              m.column = originalPosition.column
            }
            if (
              originalPosition.line !== null &&
              originalPosition.column !== null
            ) {
              m.code = generateCodeFrame(originalPosition.sourceContent, {
                line: originalPosition.line,
                column: originalPosition.column,
              }).replace(/\t/g, ' ')
            }
          }
        }
      }
      const msg = options.format(m)
      if (msg) {
        msgs.push(msg)
      }
    }
  }
  return msgs.join('\n')
}
