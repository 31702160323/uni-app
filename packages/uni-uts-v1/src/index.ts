import { isArray } from '@vue/shared'

import { runKotlinProd, runKotlinDev } from './kotlin'
import { runSwiftProd, runSwiftDev } from './swift'

import { genProxyCode, resolvePlatformIndex, resolveRootIndex } from './code'
import { resolvePackage } from './utils'

export * from './sourceMap'

export async function compile(module: string) {
  const pkg = resolvePackage(module)
  if (!pkg) {
    return
  }
  const deps: string[] = []
  const code = await genProxyCode(module, pkg)
  if (process.env.NODE_ENV !== 'development') {
    // 生产模式 支持同时生成 android 和 ios 的 uts 插件
    if (
      process.env.UNI_UTS_PLATFORM === 'app-android' ||
      process.env.UNI_UTS_PLATFORM === 'app'
    ) {
      const filename =
        resolvePlatformIndex('app-android', module, pkg) ||
        resolveRootIndex(module, pkg)
      if (filename) {
        await getCompiler('kotlin').runProd(filename)
      }
    }
    if (
      process.env.UNI_UTS_PLATFORM === 'app-ios' ||
      process.env.UNI_UTS_PLATFORM === 'app'
    ) {
      const filename =
        resolvePlatformIndex('app-ios', module, pkg) ||
        resolveRootIndex(module, pkg)
      if (filename) {
        await getCompiler('swift').runProd(filename)
      }
    }
  } else {
    if (
      process.env.UNI_UTS_PLATFORM === 'app-android' ||
      process.env.UNI_UTS_PLATFORM === 'app-ios'
    ) {
      // dev 模式
      const filename =
        resolvePlatformIndex(process.env.UNI_UTS_PLATFORM, module, pkg) ||
        resolveRootIndex(module, pkg)
      const compilerType =
        process.env.UNI_UTS_PLATFORM === 'app-android' ? 'kotlin' : 'swift'

      if (filename) {
        deps.push(filename)
        const res = await getCompiler(compilerType).runDev(filename)
        if (res) {
          if (isArray(res.deps) && res.deps.length) {
            // 添加其他文件的依赖
            deps.push(...res.deps)
          }
          if (res.type === 'kotlin') {
            const files: string[] = []
            if (process.env.UNI_APP_CHANGED_DEX_FILES) {
              try {
                files.push(...JSON.parse(process.env.UNI_APP_CHANGED_DEX_FILES))
              } catch (e) {}
            }
            if (res.dex) {
              files.push(res.dex)
            }
            process.env.UNI_APP_CHANGED_DEX_FILES = JSON.stringify([
              ...new Set(files),
            ])
          } else if (res.type === 'swift') {
            if (code) {
              throw res.msg
            }
          }
        }
      }
    }
  }
  return {
    code,
    deps,
  }
}

function getCompiler(type: 'kotlin' | 'swift') {
  if (type === 'swift') {
    return {
      runProd: runSwiftProd,
      runDev: runSwiftDev,
    }
  }
  return {
    runProd: runKotlinProd,
    runDev: runKotlinDev,
  }
}
