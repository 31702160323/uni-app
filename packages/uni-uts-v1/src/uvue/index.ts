import path from 'path'
import fs from 'fs-extra'
import {
  D8_DEFAULT_ARGS,
  KotlinCompilerServer,
  RunKotlinDevResult,
  getUniModulesCacheJars,
  getUniModulesJars,
  resolveKotlincArgs,
  createStderrListener,
} from '../kotlin'
import { parseUTSSyntaxError } from '../stacktrace'
import {
  getCompilerServer,
  getUTSCompiler,
  resolveSourceMapFile,
  resolveUTSSourceMapPath,
} from '../utils'
import { UTSBundleOptions, UTSInputOptions, UTSResult } from '@dcloudio/uts'
import { resolveSourceMapPath } from '../shared'

const DEFAULT_IMPORTS = [
  'kotlinx.coroutines.async',
  'kotlinx.coroutines.CoroutineScope',
  'kotlinx.coroutines.Deferred',
  'kotlinx.coroutines.Dispatchers',
  'io.dcloud.uts.Map',
  'io.dcloud.uts.Set',
  'io.dcloud.uts.UTSAndroid',
  'io.dcloud.uts.*',
  'io.dcloud.uts.framework.*',
  'io.dcloud.uts.vue.*',
  'io.dcloud.uts.vue.shared.*',
  'io.dcloud.uniapp.runtime.*',
]

export interface CompileAppOptions {
  inputDir: string
  outputDir: string
  package: string
  sourceMap: boolean
  uni_modules: string[]
  extApis?: Record<string, [string, string]>
  split?: boolean
  disableSplitManifest?: boolean
}
export async function compileApp(entry: string, options: CompileAppOptions) {
  const split = !!options.split
  const { bundle, UTSTarget } = getUTSCompiler()
  const imports = [...DEFAULT_IMPORTS]
  const isProd = process.env.NODE_ENV !== 'development'
  const {
    package: pkg,
    inputDir,
    outputDir,
    sourceMap,
    uni_modules,
    extApis,
  } = options

  const input: UTSInputOptions = {
    root: inputDir,
    filename: entry,
    paths: {
      vue: 'io.dcloud.uts.vue',
    },
    uniModules: uni_modules,
    globals: {
      envs: {
        // 自动化测试
        NODE_ENV: process.env.NODE_ENV,
        UNI_AUTOMATOR_WS_ENDPOINT: process.env.UNI_AUTOMATOR_WS_ENDPOINT || '',
      },
    },
  }

  const bundleOptions: UTSBundleOptions = {
    input,
    output: {
      isX: true,
      isPlugin: false,
      outDir: isProd
        ? kotlinSrcDir(path.resolve(outputDir, '.uniappx/android/'))
        : kotlinSrcDir(kotlinDir(outputDir)),
      package: pkg,
      sourceMap: sourceMap !== false ? resolveUTSSourceMapPath() : false,
      extname: 'kt',
      imports,
      logFilename: true,
      noColor: true,
      split,
      disableSplitManifest: options.disableSplitManifest,
      transform: {
        uniExtApiDefaultNamespace: 'io.dcloud.uts.extapi',
        uniExtApiNamespaces: extApis,
        uvueClassNamePrefix: 'Gen',
      },
    },
  }
  // const time = Date.now()
  // console.log(bundleOptions)
  const result = await bundle(UTSTarget.KOTLIN, bundleOptions)
  // console.log('UTS编译耗时: ' + (Date.now() - time) + 'ms')
  if (!result) {
    return
  }

  if (result.error) {
    throw parseUTSSyntaxError(result.error, inputDir)
  }

  if (isProd) {
    return runKotlinBuild(options, result)
  }

  return runKotlinDev(options, result as RunKotlinDevResult)
}

function kotlinDir(outputDir: string) {
  return path.resolve(outputDir, '../.kotlin')
}

function kotlinSrcDir(kotlinDir: string) {
  return path.resolve(kotlinDir, 'src')
}

function kotlinClassDir(kotlinDir: string) {
  return path.resolve(kotlinDir, 'class')
}

async function runKotlinDev(
  options: CompileAppOptions,
  result: RunKotlinDevResult
) {
  result.type = 'kotlin'
  const kotlinRootOutDir = kotlinDir(options.outputDir)
  const kotlinSrcOutDir = kotlinSrcDir(kotlinRootOutDir)
  const kotlinChangedFiles = result.changed.map((file) => {
    return path.resolve(kotlinSrcOutDir, file)
  })

  const { inputDir, outputDir } = options
  const kotlinMainFile = path.resolve(kotlinSrcOutDir, result.filename!)
  // 开发模式下，需要生成 dex
  if (kotlinChangedFiles.length && fs.existsSync(kotlinMainFile)) {
    const compilerServer = getCompilerServer<KotlinCompilerServer>(
      'uniapp-runextension'
    )
    if (!compilerServer) {
      throw `项目使用了uts插件，正在安装 uts Android 运行扩展...`
    }
    const {
      getDefaultJar,
      getKotlincHome,
      compile: compileDex,
    } = compilerServer

    const cacheDir = process.env.HX_DEPENDENCIES_DIR || ''

    const kotlinClassOutDir = kotlinClassDir(kotlinRootOutDir)
    const waiting = { done: undefined }
    const options = {
      version: 'v2',
      kotlinc: resolveKotlincArgs(
        kotlinChangedFiles,
        kotlinClassOutDir,
        getKotlincHome(),
        [kotlinClassOutDir].concat(
          getDefaultJar(2)
            .concat(getUniModulesCacheJars(cacheDir))
            .concat(getUniModulesJars(outputDir))
        )
      ).concat(['-module-name', `main-${+Date.now()}`]),
      d8: D8_DEFAULT_ARGS,
      kotlinOutDir: kotlinClassOutDir,
      dexOutDir: outputDir,
      inputDir: kotlinSrcOutDir,
      sourceRoot: inputDir,
      sourceMapPath: resolveSourceMapFile(outputDir, kotlinMainFile),
      stderrListener: createStderrListener(
        kotlinSrcOutDir,
        resolveSourceMapPath(),
        waiting
      ),
    }
    // console.log('DEX编译参数:', options)
    const { code, msg, data } = await compileDex(options, inputDir)
    // 等待 stderrListener 执行完毕
    if (waiting.done) {
      await waiting.done
    }
    // console.log('DEX编译结果:', code, data)
    if (!code && data) {
      result.changed = data.dexList
    } else {
      result.changed = []
      if (msg) {
        console.error(msg)
      }
    }
  }
  return result
}

async function runKotlinBuild(_options: CompileAppOptions, _result: UTSResult) {
  // TODO
}
