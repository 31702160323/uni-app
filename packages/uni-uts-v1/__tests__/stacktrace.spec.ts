import path from 'path'
import { parseUtsSwiftPluginStacktrace } from '../src/stacktrace'
const stacktrace = `/uts-development-ios/dependences/buildFramework/template/xcode_ust_template/unimoduleTestUts1/src/index.swift:3:12: error: cannot convert return expression of type 'Int' to return type 'String'\n    return 1;\n           ^\nnote: Building targets in dependency order\n/Applications/HBuilderX-Alpha.app/Contents/HBuilderX/plugins/uts-development-ios/dependences/buildFramework/template/xcode_ust_template/unimoduleTestUts1/UTS.xcodeproj: warning: The iOS deployment target 'IPHONEOS_DEPLOYMENT_TARGET' is set to 9.0, but the range of supported deployment target versions is 11.0 to 16.1.99. (in target 'unimoduleTestUts1' from project 'UTS')\n`

describe('uts:stacktrace', () => {
  test('parseUtsSwiftPluginStacktrace', async () => {
    const codes = await parseUtsSwiftPluginStacktrace({
      stacktrace,
      sourceMapFile: path.resolve(
        __dirname,
        './examples/sourcemap/index.swift.map'
      ),
      sourceRoot: '/Users/fxy/DCloud/test-uts',
    })
    expect(codes).toContain(
      `uni_modules/test-uts1/utssdk/app-ios/index.uts:2:10`
    )
  })
})
