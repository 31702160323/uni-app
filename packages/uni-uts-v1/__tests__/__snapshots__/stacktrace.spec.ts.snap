// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uts:stacktrace parseUTSKotlinStacktrace 1`] = `
"‌error: Unresolved reference: test‌
12 |          methods: {
13 |              onClick() {
14 |                  console.log(test)
   |                              ^
15 |              }
16 |          }
at components/test/test.vue:14:28"
`;

exports[`uts:stacktrace parseUTSKotlinStacktrace 2`] = `
"‌error: Unresolved reference: test‌
2  |      export default {
3  |          onLaunch: function() {
4  |              console.log('App Launch', test)
   |                                        ^
5  |          },
6  |          onShow: function() {
at App.vue:4:38"
`;
